fl4health/clients/fenda_client.py

```diff
@@ -110,8 +124,67 @@ def update_before_train(self, current_server_round: int) -> None:
         assert isinstance(self.model, FendaModel)
         if self.perfcl_loss_weights:
             self.aggregated_global_module = self.clone_and_freeze_model(self.model.global_module)
+        if self.mkmmd_loss_weights and self.old_global_module and self.aggregated_global_module:
+            local_distribution, old_global_distribution, aggregated_distribution = self.update_buffers(
+                local_module=self.model.local_module,
+                global_module=self.old_global_module,
+                aggregated_module=self.aggregated_global_module,
+            )
+            if self.mkmmd_loss_weights[0] != 0.0:
+                self.mkmmd_loss_min.betas = self.mkmmd_loss_min.optimize_betas(
+                    X=old_global_distribution, Y=aggregated_distribution, lambda_m=1e-5
+                )
+            if self.mkmmd_loss_weights[1] != 0.0:
+                self.mkmmd_loss_max.betas = self.mkmmd_loss_max.optimize_betas(
+                    X=local_distribution, Y=aggregated_distribution, lambda_m=1e-5
+                )
+
         return super().update_before_train(current_server_round)
 
+    def update_buffers(
+        self, local_module: torch.nn.Module, global_module: torch.nn.Module, aggregated_module: torch.nn.Module
+    ) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Update the feature buffer of the local, global and aggregated features."""
+        assert isinstance(local_module, torch.nn.Module)
+        assert isinstance(global_module, torch.nn.Module)
+        assert isinstance(aggregated_module, torch.nn.Module)
+
+        local_buffer = []
+        global_buffer = []
+        aggregated_buffer = []
+
+        # Save the state of the local and global modules
+        init_state_local_module = local_module.training
+        init_state_global_module = global_module.training
+
+        # Compute the old features before aggregation and global features
+        local_module.eval()
+        global_module.eval()
+        aggregated_module.eval()
+        assert not local_module.training
+        assert not global_module.training
+        assert not aggregated_module.training
+
+        with torch.no_grad():
+            for input, target in self.train_loader:
+                input, target = input.to(self.device), target.to(self.device)
+                local_features = local_module.forward(input)
+                global_features = global_module.forward(input)
+                aggregated_features = aggregated_module.forward(input)
+
+                # Local feature are same as old local features
+                local_buffer.append(local_features.reshape(len(local_features), -1))
```

Comment:
To make the code a bit more readable/re-usable, maybe we can consolidate this flattening into a private function like `def _flatten_features` and add a comment that we're assuming batch first.

Reply:
Where should we define such a function? in utils or separately for each client?

Reply:
Yeah that's a great question. This functionality is used several other classes too. I'd say a utils function for now I guess. I don't love just throwing functions randomly into a "catch-all" folder, but I think that makes the most sense as a quick thing. Maybe put a #TODO above the function to find a better place for it.